torch_glow/src/PyTorchFileLoader.cpp

```diff
@@ -15,30 +15,189 @@
+}
+
+/// Runs Module forward pass, triggers custom fusion pass if local thread
+/// Glow function is set.
+llvm::Error
+evaluateModuleGraph(std::shared_ptr<torch::jit::script::Module> &module,
+                    std::vector<torch::jit::IValue> &inputs) {
+  try {
+    module->forward(inputs);
+  } catch (const std::exception &x) {
+    RETURN_ERR(x.what());
+  }
+  return llvm::Error::success();
+}
 
-  if (errPtr) {
-    *errPtr = setup();
-  } else {
-    EXIT_ON_ERR(setup());
+/// Helper struct, which on constructor registers custom fusion pass
+/// and operator.
+struct RegisterCustomFusionPass {
+  RegisterCustomFusionPass() {
+    auto options = c10::OperatorOptions();
+    options.setAliasAnalysis(at::AliasAnalysisKind::PURE);
+    torch::jit::RegisterOperators op({torch::jit::Operator(
+        getFusionSymbol(),
+        [](const torch::jit::Node *node) {
+          return [node](torch::jit::Stack &stack) {
+            // Get JIT Graph.
+            auto graph = node->g(at::attr::Subgraph);
+            auto err = loadJitGraphToGlowFunction(
+                stack, *graph, *localFusionInfo.function,
+                *localFusionInfo.inputPlaceholders,
+                *localFusionInfo.outputPlaceholders, *localFusionInfo.settings);
+            if (static_cast<bool>(err)) {
+              throw std::invalid_argument(llvm::toString(std::move(err)));
+            }
+            return 0;
+          };
+        },
+        options)});
+
+    torch::jit::RegisterPass pass([&](std::shared_ptr<torch::jit::Graph> &g) {
+      // Trigger custom fusion pass only if local thread Glow Function is set.
+      if (localFusionInfo.function) {
+        glowCustomFuse(g, getFusionSymbol());
+      }
+    });
+  }
+};
+} // namespace
+
+/*static*/
+llvm::Error PyTorchFileLoader::loadPyTorchModel(
+    const std::string &fileName,
+    std::shared_ptr<torch::jit::script::Module> &module) {
+  try {
+    module = std::make_shared<torch::jit::script::Module>(
+        torch::jit::load(fileName));
+  } catch (const std::exception &x) {
+    RETURN_ERR(strFormat("Cannot load model from file: %s, , reason: %s",
+                         fileName.c_str(), x.what()));
+  }
+  return llvm::Error::success();
+}
+
+/*static*/
+llvm::Error PyTorchFileLoader::loadPyTorchGraph(
+    const std::string &fileName, std::vector<torch::jit::IValue> &inputs,
+    glow::Function &F, std::vector<glow::Placeholder *> &inputPlaceholders,
+    std::vector<glow::Placeholder *> &outputPlaceholders,
+    const PyTorchLoaderSettings &settings, bool sanityCheck) {
+  // Register custom pass once.
+  static RegisterCustomFusionPass fuser;
+
+  // Convert PyTorch model into JIT Module.
+  std::shared_ptr<torch::jit::script::Module> module;
+  RETURN_IF_ERR(PyTorchFileLoader::loadPyTorchModel(fileName, module));
+
+  // Disable gradient nodes generation.
+  at::NoGradGuard guard;
+
+  // Set thread local pointer, not null values activate custom fusion pass.
```

Comment:
doxygen

Reply:
can you explain, what should be changed to comply with doxygen style?.